File: project/src/banksystem/BankSystem.java
package banksystem;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Scanner;

import banksystem.helper.InputHelper;
import banksystem.helper.ValidationHelper;
import banksystem.manager.AccountManager;
import banksystem.manager.AdminManager;
import banksystem.manager.SchedulerManager;
import banksystem.manager.TransactionManager;
import banksystem.manager.UserManager;

public class BankSystem {
	private Scanner scanner = new Scanner(System.in);
	private Connection conn = null;
	private String loginId = null;
	private String adminLoginId = null;

	// helper 및 manager 객체들
	private ValidationHelper validator;
	private InputHelper inputHelper;
	private UserManager userManager;
	private AccountManager accountManager;
	private TransactionManager transactionManager;
	private AdminManager adminManager;
	private SchedulerManager schedulerManager;

	public BankSystem() {
		try {
			Class.forName("oracle.jdbc.OracleDriver");
			conn = DriverManager.getConnection("jdbc:oracle:thin:@localhost:1521/orcl", "jhw1", "1234");
			conn.setAutoCommit(true);
			System.out.println("은행 계좌 시스템 DB 연결 성공!" + "\n");

			// 기존 manager 객체 초기화
			validator = new ValidationHelper(conn);
			inputHelper = new InputHelper(scanner, validator, conn);
			userManager = new UserManager(conn, validator, inputHelper, scanner);

			transactionManager = new TransactionManager(conn, inputHelper, null, scanner);
			accountManager = new AccountManager(conn, validator, inputHelper, userManager, scanner, transactionManager);
			setAccountManagerToTransactionManager();
			inputHelper.setAccountManager(accountManager);
			adminManager = new AdminManager(conn, validator, inputHelper, scanner);

			// 스케줄러 초기화 및 시작
			schedulerManager = new SchedulerManager(conn, adminManager);
			schedulerManager.start();

			// 시스템 종료 시 스케줄러도 함께 정리
			Runtime.getRuntime().addShutdownHook(new Thread(() -> {
				System.out.println("\n시스템이 종료됩니다...");
				if (schedulerManager != null) {
					schedulerManager.stop();
				}
				System.out.println("안전하게 종료되었습니다.");
			}));

		} catch (Exception e) {
			e.printStackTrace();
			exit();
		}
	}

	private void setAccountManagerToTransactionManager() {
		if (transactionManager != null && accountManager != null) {
			transactionManager.setAccountManager(accountManager);
		}
	}

	private void exit() {
		try {
			// 종료 시 스케줄러 정리
			if (schedulerManager != null) {
				schedulerManager.stop();
			}

			if (conn != null)
				conn.close();
		} catch (SQLException e) {
			System.out.println("DB 연결 종료 중 오류: " + e.getMessage());
		}
		System.out.println("은행 시스템이 정상적으로 종료되었습니다.");
		System.exit(0);
	}

	private void logout() {
		loginId = null;
		adminLoginId = null;
		System.out.println("로그아웃되었습니다.");
		list();
	}

	private void list() {
		if (loginId == null && adminLoginId == null) {
			System.out.println("\n-- 은행 계좌 관리 시스템 --");
			System.out.println("계좌 서비스를 이용하려면 로그인해주세요.");

			// 스케줄러 상태 표시
			if (schedulerManager != null && schedulerManager.isRunning()) {
				System.out.println("✅ 자동 이자 지급: " + schedulerManager.getNextExecutionInfo());
			}

			menu();
			return;
		}

		if (adminLoginId != null) {
			System.out.println("\n[관리자 모드] " + adminManager.getAdminName(adminLoginId) + "님");

			// 스케줄러 상태만 표시
			if (schedulerManager != null) {
				System.out.println(schedulerManager.getNextExecutionInfo());
			}

		} else {
			accountManager.listAccounts(loginId);
		}
		menu();
	}

	private void menu() {
		System.out.println(
				"============================================================================================================================");

		if (loginId == null && adminLoginId == null) {
			System.out.println("✅ 메인메뉴: 1.회원가입 | 2.사용자 로그인 | 3.관리자 로그인 | 4.종료");
			System.out.print("메뉴선택: ");

			String menuNo = scanner.nextLine();
			switch (menuNo) {
			case "1" -> {
				userManager.join();
				list();
			}
			case "2" -> {
				String userId = userManager.login();
				if (userId != null) {
					loginId = userId;
				}
				list();
			}
			case "3" -> {
				String adminId = adminManager.adminLogin();
				if (adminId != null) {
					adminLoginId = adminId;
				}
				list();
			}
			case "4" -> exit();
			default -> {
				System.out.println("1 ~ 4번의 숫자만 입력이 가능합니다.");
				menu();
			}
			}

		} else if (adminLoginId != null) {
			// 관리자 메뉴에 스케줄러 관리 추가
			System.out.println("✅ 계좌관리\t\t✅ 이자관리\t\t✅ 시스템관리\t\t✅ 기타");
			System.out.println(
					"============================================================================================================================");
			System.out.println("1. 전체계좌조회\t\t3. 수동이자지급\t\t5. 스케줄러상태\t\t6. 로그아웃");
			System.out.println("2. 사용자별계좌조회\t\t4. 이자지급내역조회\t\t0. 종료");
			System.out.println("============================================================================================================================");
			System.out.print("메뉴선택: ");

			String menuNo = scanner.nextLine();
			switch (menuNo) {
			case "1" -> {
				adminManager.viewAllAccounts();
				list();
			}
			case "2" -> {
				adminManager.viewUserAccounts();
				list();
			}
			case "3" -> { 
				executeManualInterestPayment();
				list();
			}
			case "4" -> {
				adminManager.viewInterestHistory();
				list();
			}
			case "5" -> { 
				showSchedulerStatus();
				list();
			}
			case "6" -> logout();
			case "0" -> exit();
			default -> {
				System.out.println("0 ~ 6번의 숫자만 입력이 가능합니다.");
				menu();
			}
			}

		} else {
			System.out.println("✅ 계좌관리\t✅ 거래업무\t✅ 기타설정\t✅ 시스템");
			System.out.println(
					"============================================================================================================================");
			System.out.println("1. 계좌생성\t4. 입금\t\t8. 계좌비밀번호변경\t10. 로그아웃");
			System.out.println("2. 계좌조회\t5. 출금\t\t9. 회원정보수정\t0. 종료");
			System.out.println("3. 계좌해지\t6. 이체");
			System.out.println("\t\t7. 거래내역조회");
			System.out.println("============================================================================================================================");
			System.out.print("메뉴선택: ");

			String menuNo = scanner.nextLine();
			switch (menuNo) {
			case "1" -> {
				accountManager.createAccount(loginId);
				list();
			}
			case "2" -> {
				accountManager.readAccount(loginId);
				list();
			}
			case "3" -> {
				accountManager.deleteAccountMenu(loginId, null);
				list();
			}
			case "4" -> {
				transactionManager.deposit(loginId);
				list();
			}
			case "5" -> {
				transactionManager.withdraw(loginId);
				list();
			}
			case "6" -> {
				transactionManager.transfer(loginId);
				list();
			}
			case "7" -> {
				transactionManager.history(loginId);
				menu();
			}
			case "8" -> {
				accountManager.changePassword(loginId);
				list();
			}
			case "9" -> {
				userManager.modifyUserInfo(loginId);
				list();
			}
			case "10" -> logout();
			case "0" -> exit();
			default -> {
				System.out.println("0 ~ 10번의 숫자만 입력이 가능합니다.");
				menu();
			}
			}
		}
	}

	// 스케줄러 상태 표시
	private void showSchedulerStatus() {
		System.out.println("\n[스케줄러 상태 정보]");
		System.out.println(
				"============================================================================================================================");

		if (schedulerManager != null) {
			if (schedulerManager.isRunning()) {
				System.out.println("상태: 실행 중");
				System.out.println(schedulerManager.getNextExecutionInfo());
				System.out.println("실행 시간: 매월 마지막 날 오후 2시");
				System.out.println("실행 내용: 모든 계좌 이자 일괄 지급");
			} else {
				System.out.println("상태: 중지됨");
			}
		} else {
			System.out.println("상태: 스케줄러가 초기화되지 않음");
		}

		System.out.println("============================================================================================================================");
		System.out.println("엔터키를 누르면 메뉴로 돌아갑니다.");
		scanner.nextLine();
	}

	// 수동 이자 지급 
	private void executeManualInterestPayment() {
		System.out.println("\n[수동 이자 지급]");
		System.out.println(
				"============================================================================================================================");
		System.out.println("   주의사항:");
		System.out.println("   - 자동 스케줄러가 현재 실행 중입니다.");

		if (schedulerManager != null && schedulerManager.isRunning()) {
			System.out.println("   - " + schedulerManager.getNextExecutionInfo());
		}

		System.out.println("   - 수동 지급은 긴급상황이나 테스트 목적으로만 사용하세요.");
		System.out.println("   - 정기 이자 지급은 자동 스케줄러가 처리합니다.");
		System.out.println(
				"============================================================================================================================");

		System.out.println("\n수동 이자 지급을 실행하는 이유를 선택하세요:");
		System.out.println("1. 긴급 상황 (고객 민원, 시스템 오류 등)");
		System.out.println("2. 테스트 목적");
		System.out.println("3. 특별 지급 (이벤트, 보상 등)");
		System.out.println("4. 스케줄러 문제로 인한 수동 실행");
		System.out.println("5. 취소");
		System.out.print("선택 (1 - 5): ");

		String choice = scanner.nextLine();
		String reason = "";

		switch (choice) {
		case "1" -> reason = "MANUAL_EMERGENCY";
		case "2" -> reason = "MANUAL_TEST";
		case "3" -> reason = "MANUAL_SPECIAL";
		case "4" -> reason = "MANUAL_SCHEDULER_ISSUE";
		case "5" -> {
			System.out.println("수동 이자 지급이 취소되었습니다.");
			return;
		}
		default -> {
			System.out.println("잘못된 선택입니다. 취소됩니다.");
			return;
		}
		}

		System.out.println("\n정말로 수동 이자 지급을 실행하시겠습니까?");
		if (inputHelper.confirmAction()) {
			System.out.println("수동 이자 지급을 실행합니다... (사유: " + reason + ")");
			adminManager.executeInterestPayment(adminLoginId + "_" + reason);
			System.out.println("✅ 수동 이자 지급이 완료되었습니다!");
		} else {
			System.out.println("수동 이자 지급이 취소되었습니다.");
		}
	}

	public static void main(String[] args) {
		BankSystem bankSystem = new BankSystem();
		bankSystem.list();
	}
}

File: 4day/bank_system/bank_system.py
"""
은행 시스템 메인 클래스
Java의 BankSystem 클래스를 Python으로 변환
"""

import sys
import signal
from typing import Optional
from .database import get_database_connection, close_database_connection
from .managers.user_manager import UserManager
from .managers.account_manager import AccountManager
from .managers.transaction_manager import TransactionManager
from .managers.admin_manager import AdminManager
from .managers.scheduler_manager import SchedulerManager


class BankSystem:
    """은행 시스템의 메인 클래스"""
    
    def __init__(self):
        """BankSystem 초기화"""
        self.login_id: Optional[str] = None
        self.admin_login_id: Optional[str] = None
        
        # 데이터베이스 연결
        try:
            self.db = get_database_connection()
            print("은행 계좌 시스템 DB 연결 성공!\n")
        except Exception as e:
            print(f"데이터베이스 연결 실패: {e}")
            self.exit()
            return
        
        # 매니저 객체들 초기화
        try:
            self.user_manager = UserManager()
            self.transaction_manager = TransactionManager()
            self.account_manager = AccountManager(
                user_manager=self.user_manager,
                transaction_manager=self.transaction_manager
            )
            self.transaction_manager.set_account_manager(self.account_manager)
            self.admin_manager = AdminManager()
            
            # 스케줄러 초기화 및 시작
            self.scheduler_manager = SchedulerManager(self.admin_manager)
            self.scheduler_manager.start()
            
            # 시스템 종료 시 정리 작업
            self._setup_shutdown_handler()
            
        except Exception as e:
            print(f"시스템 초기화 오류: {e}")
            self.exit()
    
    def _setup_shutdown_handler(self):
        """시스템 종료 시 정리 작업 설정"""
        def signal_handler(signum, frame):
            print("\n시스템이 종료됩니다...")
            if self.scheduler_manager:
                self.scheduler_manager.stop()
            print("안전하게 종료되었습니다.")
            sys.exit(0)
        
        signal.signal(signal.SIGINT, signal_handler)
        signal.signal(signal.SIGTERM, signal_handler)
    
    def run(self):
        """시스템 실행"""
        try:
            self.list()
        except KeyboardInterrupt:
            print("\n시스템이 종료됩니다...")
            self.exit()
        except Exception as e:
            print(f"시스템 실행 오류: {e}")
            self.exit()
    
    def exit(self):
        """시스템 종료"""
        try:
            # 스케줄러 정리
            if hasattr(self, 'scheduler_manager') and self.scheduler_manager:
                self.scheduler_manager.stop()
            
            # 데이터베이스 연결 종료
            close_database_connection()
            
        except Exception as e:
            print(f"시스템 종료 중 오류: {e}")
        
        print("은행 시스템이 정상적으로 종료되었습니다.")
        sys.exit(0)
    
    def logout(self):
        """로그아웃"""
        self.login_id = None
        self.admin_login_id = None
        print("로그아웃되었습니다.")
        self.list()
    
    def list(self):
        """메인 화면 표시"""
        if self.login_id is None and self.admin_login_id is None:
            print("\n-- 은행 계좌 관리 시스템 --")
            print("계좌 서비스를 이용하려면 로그인해주세요.")
            
            # 스케줄러 상태 표시
            if self.scheduler_manager and self.scheduler_manager.is_running():
                print(f"✅ 자동 이자 지급: {self.scheduler_manager.get_next_execution_info()}")
            
            self.menu()
            return
        
        if self.admin_login_id is not None:
            print(f"\n[관리자 모드] {self.admin_manager.get_admin_name(self.admin_login_id)}님")
            
            # 스케줄러 상태 표시
            if self.scheduler_manager:
                print(self.scheduler_manager.get_next_execution_info())
        else:
            self.account_manager.list_accounts(self.login_id)
        
        self.menu()
    
    def menu(self):
        """메뉴 표시 및 처리"""
        print("=" * 120)
        
        if self.login_id is None and self.admin_login_id is None:
            # 메인 메뉴
            print("✅ 메인메뉴: 1.회원가입 | 2.사용자 로그인 | 3.관리자 로그인 | 4.종료")
            choice = input("메뉴선택: ").strip()
            
            if choice == "1":
                self.user_manager.join()
                self.list()
            elif choice == "2":
                user_id = self.user_manager.login()
                if user_id:
                    self.login_id = user_id
                self.list()
            elif choice == "3":
                admin_id = self.admin_manager.admin_login()
                if admin_id:
                    self.admin_login_id = admin_id
                self.list()
            elif choice == "4":
                self.exit()
            else:
                print("1 ~ 4번의 숫자만 입력이 가능합니다.")
                self.menu()
        
        elif self.admin_login_id is not None:
            # 관리자 메뉴
            print("✅ 계좌관리\t\t✅ 이자관리\t\t✅ 시스템관리\t\t✅ 기타")
            print("=" * 120)
            print("1. 전체계좌조회\t\t3. 수동이자지급\t\t5. 스케줄러상태\t\t6. 로그아웃")
            print("2. 사용자별계좌조회\t\t4. 이자지급내역조회\t\t0. 종료")
            print("=" * 120)
            choice = input("메뉴선택: ").strip()
            
            if choice == "1":
                self.admin_manager.view_all_accounts()
                self.list()
            elif choice == "2":
                self.admin_manager.view_user_accounts()
                self.list()
            elif choice == "3":
                self.execute_manual_interest_payment()
                self.list()
            elif choice == "4":
                self.admin_manager.view_interest_history()
                self.list()
            elif choice == "5":
                self.show_scheduler_status()
                self.list()
            elif choice == "6":
                self.logout()
            elif choice == "0":
                self.exit()
            else:
                print("0 ~ 6번의 숫자만 입력이 가능합니다.")
                self.menu()
        
        else:
            # 사용자 메뉴
            print("✅ 계좌관리\t✅ 거래업무\t✅ 기타설정\t✅ 시스템")
            print("=" * 120)
            print("1. 계좌생성\t4. 입금\t\t8. 계좌비밀번호변경\t10. 로그아웃")
            print("2. 계좌조회\t5. 출금\t\t9. 회원정보수정\t0. 종료")
            print("3. 계좌해지\t6. 이체")
            print("\t\t7. 거래내역조회")
            print("=" * 120)
            choice = input("메뉴선택: ").strip()
            
            if choice == "1":
                self.account_manager.create_account(self.login_id)
                self.list()
            elif choice == "2":
                self.account_manager.read_account(self.login_id)
                self.list()
            elif choice == "3":
                self.account_manager.delete_account_menu(self.login_id)
                self.list()
            elif choice == "4":
                self.transaction_manager.deposit(self.login_id)
                self.list()
            elif choice == "5":
                self.transaction_manager.withdraw(self.login_id)
                self.list()
            elif choice == "6":
                self.transaction_manager.transfer(self.login_id)
                self.list()
            elif choice == "7":
                self.transaction_manager.history(self.login_id)
                self.menu()
            elif choice == "8":
                self.account_manager.change_password(self.login_id)
                self.list()
            elif choice == "9":
                self.user_manager.modify_user_info(self.login_id)
                self.list()
            elif choice == "10":
                self.logout()
            elif choice == "0":
                self.exit()
            else:
                print("0 ~ 10번의 숫자만 입력이 가능합니다.")
                self.menu()
    
    def show_scheduler_status(self):
        """스케줄러 상태 표시"""
        print("\n[스케줄러 상태 정보]")
        print("=" * 120)
        
        if self.scheduler_manager:
            if self.scheduler_manager.is_running():
                print("상태: 실행 중")
                print(self.scheduler_manager.get_next_execution_info())
                print("실행 시간: 매월 마지막 날 오후 2시")
                print("실행 내용: 모든 계좌 이자 일괄 지급")
            else:
                print("상태: 중지됨")
        else:
            print("상태: 스케줄러가 초기화되지 않음")
        
        print("=" * 120)
        input("엔터키를 누르면 메뉴로 돌아갑니다.")
    
    def execute_manual_interest_payment(self):
        """수동 이자 지급"""
        print("\n[수동 이자 지급]")
        print("=" * 120)
        print("   주의사항:")
        print("   - 자동 스케줄러가 현재 실행 중입니다.")
        
        if self.scheduler_manager and self.scheduler_manager.is_running():
            print(f"   - {self.scheduler_manager.get_next_execution_info()}")
        
        print("   - 수동 지급은 긴급상황이나 테스트 목적으로만 사용하세요.")
        print("   - 정기 이자 지급은 자동 스케줄러가 처리합니다.")
        print("=" * 120)
        
        print("\n수동 이자 지급을 실행하는 이유를 선택하세요:")
        print("1. 긴급 상황 (고객 민원, 시스템 오류 등)")
        print("2. 테스트 목적")
        print("3. 특별 지급 (이벤트, 보상 등)")
        print("4. 스케줄러 문제로 인한 수동 실행")
        print("5. 취소")
        choice = input("선택 (1 - 5): ").strip()
        
        reason_map = {
            "1": "MANUAL_EMERGENCY",
            "2": "MANUAL_TEST", 
            "3": "MANUAL_SPECIAL",
            "4": "MANUAL_SCHEDULER_ISSUE"
        }
        
        if choice == "5":
            print("수동 이자 지급이 취소되었습니다.")
            return
        
        if choice not in reason_map:
            print("잘못된 선택입니다. 취소됩니다.")
            return
        
        reason = reason_map[choice]
        
        print("\n정말로 수동 이자 지급을 실행하시겠습니까?")
        if self.user_manager.input_helper.confirm_action():
            print(f"수동 이자 지급을 실행합니다... (사유: {reason})")
            self.admin_manager.execute_interest_payment(f"{self.admin_login_id}_{reason}")
            print("✅ 수동 이자 지급이 완료되었습니다!")
        else:
            print("수동 이자 지급이 취소되었습니다.")


def main():
    """메인 함수"""
    try:
        bank_system = BankSystem()
        bank_system.run()
    except Exception as e:
        print(f"시스템 시작 오류: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
